### PR TITLE
nRF2 clock control API

### PR DESCRIPTION
**Pull request by bjarki-andreasen**
This PR continues from #76430

### CODE REVIEW

File: drivers/clock_control/clock_control_nrf2_lfclk.c [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+#define NRF2_LFCLK_MAX_ACCURACY DT_INST_PROP(0, synth_accuracy_ppm)
+#endif
+
+#define NRFS_CLOCK_TIMEOUT K_MSEC(CONFIG_CLOCK_CONTROL_NRF2_NRFS_CLOCK_TIMEOUT_MS)
+
+static const struct clock_options {
+	uint16_t accuracy : 15;
+	uint16_t precision : 1;
+	nrfs_clock_src_t src;
+} clock_options[] = {
```
Comment (hubertmis) [inline drivers/clock_control/clock_control_nrf2_lfclk.c:35]: It looks like this array is sorted by accuracy (from the worst being the highest ppm number). If that's a requirement for the algorithm, it would be good to note it somewhere, maybe in a comment?

File: drivers/clock_control/clock_control_nrf2_hsfll.c [hunk 2]
```diff
@@ -192,6 +204,28 @@ static int api_get_rate_hsfll(const struct device *dev,
...
+
+	LOG_DBG("Trim: HSFLL VSUP: 0x%.8x",   trim.vsup);
+	LOG_DBG("Trim: HSFLL COARSE: 0x%.8x", trim.coarse);
+	LOG_DBG("Trim: HSFLL FINE: 0x%.8x",   trim.fine);
+
+	nrf_hsfll_clkctrl_mult_set(dev_config->hsfll, dev_config->multiplier);
+	nrf_hsfll_trim_set(dev_config->hsfll, &trim);
+
+	nrf_hsfll_task_trigger(dev_config->hsfll, NRF_HSFLL_TASK_FREQ_CHANGE);
+	/* HSFLL task frequency change needs to be triggered twice to take effect.*/
```
Comment (hubertmis) [inline drivers/clock_control/clock_control_nrf2_hsfll.c:222]: You could refer to HMPAN-23 here for traceability.

**Review by hubertmis — COMMENTED** [review]
We should agree on dt bindings. I would like to remove redundancy from there. Other than that, it looks good to go.

File: drivers/clock_control/clock_control_nrf2_hsfll.c [hunk 3]
```diff
@@ -192,6 +204,28 @@ static int api_get_rate_hsfll(const struct device *dev,
 
 static int hsfll_init(const struct device *dev)
 {
+	const struct hsfll_dev_config *dev_config = dev->config;
+	nrf_hsfll_trim_t trim = {
+		.vsup   = sys_read32(dev_config->trim_vsup),
+		.coarse = sys_read32(dev_config->trim_coarse),
+		.fine   = sys_read32(dev_config->trim_fine),
```
Comment (hubertmis) [inline drivers/clock_control/clock_control_nrf2_hsfll.c:211]: Settint TCOEF trim is still missing. Are we aiming to fix it in this PR? If not we should create a ticket to make sure we don't forget.

File: drivers/clock_control/clock_control_nrf2_common.h [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+struct nrf2_clock_lrcconf_sink {
+	sys_snode_t node;
+};
+
+/**
+ * @brief Request or release lrcconf main power domain
+ */
+void nrf2_clock_request_lrcconf_poweron_main(struct nrf2_clock_lrcconf_sink *sink);
+void nrf2_clock_release_lrcconf_poweron_main(struct nrf2_clock_lrcconf_sink *sink);
```
[the remark this answers is not in the document]
Comment (bjarki-andreasen) [reply drivers/clock_control/clock_control_nrf2_common.h:87]: agreed, this is a temporary solution :)

File: dts/bindings/clock/nordic,nrf-fll16m.yaml [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    type: phandle
+    description: |
+      The HFXO clock is used as a clock source if the clock source is bypass, and if
+      the clock source is closed-loop and the LFXO clock is not available.
+
+  lfxo-clock:
+    type: phandle
+    description: |
+      The LFXO clock is used as a clock source if the clock source is closed-loop and
+      the LFXO clock is not available.
```
Comment (hubertmis) [inline dts/bindings/clock/nordic,nrf-fll16m.yaml:34]: typo: shoud be "LFXO clock is available"

File: dts/bindings/clock/nordic,nrf-lfclk.yaml [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    type: phandle
+    description: |
+      The HFXO clock is used as a clock source if the SYNTH clock source is selected
+      and the LFXO clock is not available.
+
+  lfxo-clock:
+    type: phandle
+    description: |
+      The LFXO clock is used as a clock source if the SYNTH clock source is selected
+      and the LFXO clock is available.
```
Comment (hubertmis) [inline dts/bindings/clock/nordic,nrf-lfclk.yaml:32]: You could add a note to both that these clocks are used indirectly through FLL16M in BYPASS mode. Something like:

The xxXO clock is used indirectly as a clock source, through FLL16M in the BYPASS mode, if the SYNTH clock source is selected...

Or maybe easier to understand:

The xxXO clock is used as a clock source if...
The xxXO is used indirectly. It is a clock source for FLL16M configured in a BYPASS mode which is a direct clock source for LFCLK in the SYNTH mode.

File: drivers/clock_control/clock_control_nrf2_hsfll.c (hunk 3 above)
Comment (bjarki-andreasen) [reply drivers/clock_control/clock_control_nrf2_hsfll.c:212]: Created a ticket, this PR just moves the existing init code from SOC to the clock driver, so it arguably is out of scope :)

File: drivers/clock_control/clock_control_nrf2_fll16m.c [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+static const struct clock_options {
+	uint16_t accuracy;
+	uint8_t mode;
+} clock_options[] = {
+	{
+		.accuracy = FLL16M_OPEN_LOOP_ACCURACY,
+		.mode = FLL16M_MODE_OPEN_LOOP,
+	},
+	{
+		/* Closed-loop mode uses LFXO as source if present, HFXO otherwise */
```
Comment (hubertmis) [inline drivers/clock_control/clock_control_nrf2_fll16m.c:52]: It seems this comment applies to lines 36-40 now

**Review by hubertmis — DISMISSED** [review]
Looks good. A little nitpicking around the doc strings and comments. But for me, it looks good to be merged and mistakes can be fixed later.

File: drivers/clock_control/clock_control_nrf2_fll16m.c (hunk 7 above)
Comment (bjarki-andreasen) [reply drivers/clock_control/clock_control_nrf2_fll16m.c]: moved to line 36

File: dts/bindings/clock/nordic,nrf-lfclk.yaml (hunk 6 above)
Comment (bjarki-andreasen) [reply dts/bindings/clock/nordic,nrf-lfclk.yaml]: updated to mention indirect use through FLL16M

File: dts/bindings/clock/nordic,nrf-fll16m.yaml (hunk 5 above)
Comment (bjarki-andreasen) [reply dts/bindings/clock/nordic,nrf-fll16m.yaml]: fixed

File: dts/bindings/clock/nordic,nrf-fll16m.yaml [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+  closed-loop-base-accuracy-ppm:
+    type: int
+    description: |
+      Base clock accuracy in parts per million if closed-loop clock source is used.
+      The actual accuracy is this property plus the accuracy of the HFXO or LFXO.
+
+  hfxo-clock:
+    type: phandle
+    description: |
+      The HFXO clock is used as a clock source if the clock source is bypass, and if
```
Comment (hubertmis) [inline dts/bindings/clock/nordic,nrf-fll16m.yaml:27]: I missed this during the previous review iteration. Looks like another typo:
```suggestion
      The HFXO clock is used as a clock source if the clock source is bypass, or if
```

**Review by hubertmis — DISMISSED** [review]

File: dts/bindings/clock/nordic,nrf-fll16m.yaml (hunk 8 above)
Comment (bjarki-andreasen) [reply dts/bindings/clock/nordic,nrf-fll16m.yaml]: done :)

**Review by hubertmis — DISMISSED** [review]
No more comments :)

File: dts/common/nordic/nrf54h20.dtsi [hunk 9]
```diff
@@ -25,20 +25,23 @@
...
+			clocks = <&cpuapp_hsfll>;
 			clock-frequency = <DT_FREQ_M(320)>;
 		};
 
 		cpurad: cpu@3 {
 			compatible = "arm,cortex-m33";
 			reg = <3>;
 			device_type = "cpu";
+			clocks = <&cpurad_hsfll>;
 			clock-frequency = <DT_FREQ_M(256)>;
```
Comment (gmarull) [inline dts/common/nordic/nrf54h20.dtsi:37]: we should at some point remove clock-frequency as it has become redundant

File: boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20-common.dtsi [hunk 10]
```diff
@@ -21,3 +21,19 @@
 	startup-time-us = <600000>;
 	mode = "crystal";
 };
+
+&fll16m {
+	status = "okay";
+	open-loop-accuracy-ppm = <20000>;
+	closed-loop-base-accuracy-ppm = <5000>;
+	hfxo-clock = <&hfxo>;
+	lfxo-clock = <&lfxo>;
```
Comment (gmarull) [inline boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20-common.dtsi:30]: why is this modified at board level?

Comment (gmarull) [reply boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20-common.dtsi:30]: also, this should be:

```
clocks = <&hfxo>, <&lfxo>;
clock-names = "hfxo", "lfxo";
```

File: boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20-common.dtsi [hunk 11]
```diff
@@ -21,3 +21,19 @@
...
+};
+
+&lfclk {
+	status = "okay";
+	lfrc-accuracy-ppm = <500>;
+	lflprc-accuracy-ppm = <1000>;
+	hfxo-clock = <&hfxo>;
+	lfxo-clock = <&lfxo>;
```
Comment (gmarull) [inline boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20-common.dtsi:38]: same

File: drivers/clock_control/clock_control_nrf2_common.c [hunk 12]
```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "clock_control_nrf2_common.h"
+#include <zephyr/logging/log.h>
+#include <zephyr/kernel.h>
+LOG_MODULE_REGISTER(clock_control_nrf2, CONFIG_CLOCK_CONTROL_LOG_LEVEL);
```
Comment (gmarull) [inline drivers/clock_control/clock_control_nrf2_common.c:9]: move after includes

File: samples/boards/nrf/nrf2_clock_control/src/main.c [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+
+	printk("request applied to clock in %llims\n", req_stop_uptime - req_start_uptime);
+	k_sleep(SAMPLE_KEEP_REQUEST_TIMEOUT);
+
+	printk("\n");
+	printk("releasing requested clock specs\n");
+	ret = nrf_clock_control_release(sample_clock_dev, &spec);
+	if (ret < 0) {
+		printk("failed to release requested clock specs\n");
+		return ret;
```
Comment (gmarull) [inline samples/boards/nrf/nrf2_clock_control/src/main.c:93]: main must always return 0

Comment (bjarki-andreasen) [reply samples/boards/nrf/nrf2_clock_control/src/main.c]: updated to return 0

File: drivers/clock_control/clock_control_nrf2_common.c (hunk 12 above)
Comment (bjarki-andreasen) [reply drivers/clock_control/clock_control_nrf2_common.c]: Moved after includes

File: boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20-common.dtsi (hunk 10 above)
Comment (bjarki-andreasen) [reply boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20-common.dtsi]: Changed to use clocks property, and moved to the common nrf54h20.dtsi

File: dts/common/nordic/nrf54h20.dtsi (hunk 9 above)
Comment (bjarki-andreasen) [reply dts/common/nordic/nrf54h20.dtsi:37]: outside the scope of this PR

**Comment by bjarki-andreasen** [comment]
Since we are now using the clocks property, the property descriptions for the properties `hfxo_clock` and `lfxo_clock` has been moved to the the descriptions of the bindings, along with a dts example for both `fll16m` and `lfclk`

**Review by gmarull — DISMISSED** [review]

**Comment by bjarki-andreasen** [comment]
missed a couple of log register macros not placed after header :)

**Review by gmarull — DISMISSED** [review]

File: drivers/clock_control/clock_control_nrf2_fll16m.c [hunk 14]
```diff
@@ -0,0 +1,263 @@
...
+#define FLL16M_MAX_ACCURACY FLL16M_HFXO_ACCURACY
+
+/* Closed-loop mode uses LFXO as source if present, HFXO otherwise */
+#if FLL16M_HAS_LFXO
+#define FLL16M_CLOSED_LOOP_ACCURACY (FLL16M_CLOSED_LOOP_BASE_ACCURACY + FLL16M_LFXO_ACCURACY)
+#else
+#define FLL16M_CLOSED_LOOP_ACCURACY (FLL16M_CLOSED_LOOP_BASE_ACCURACY + FLL16M_HFXO_ACCURACY)
+#endif
+
+/* Clock options sorted from lowest to highest accuracy */
```
Comment (hubertmis) [inline drivers/clock_control/clock_control_nrf2_fll16m.c:45]: Now when I'm thinking about this array, it actually is sorted from the lowest to the highest power consumption.
In the case of this specific array, accuracy is also improving. But it's not necessarily the rule for all similar arrays.

We aim to sort these arrays by power consumption to make the algorithm select the mode with _the lowest power consumption_, which meets the requested accuracy and precision numbers.

**Review by hubertmis — DISMISSED** [review]
When reading the PR again a few more text improvements came to my mind

**Review by gmarull — DISMISSED** [review]

File: dts/bindings/clock/nordic,nrf-fll16m.yaml [hunk 15]
```diff
@@ -0,0 +1,42 @@
+# Copyright (c) 2024 Nordic Semiconductor ASA
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+
+description: |
+  Nordic FLL16M
+
+  The FLL16M can use the following clocks as clock sources:
+
+    - HFXO: The HFXO clock is used as a clock source if the
+      clock source is bypass, or if the clock source is
```
[the remark this answers is not in the document]
Comment (bjarki-andreasen) [reply dts/bindings/clock/nordic,nrf-fll16m.yaml]: updated to use mode for the clocks FLL16M and LFCLK and clock source for the clocks HFXO and LFXO

File: drivers/clock_control/clock_control_nrf2_fll16m.c (hunk 14 above)
Comment (bjarki-andreasen) [reply drivers/clock_control/clock_control_nrf2_fll16m.c:44]: I will leave this for now, since the API clock spec requests set frequency, accuracy and precision. That's why these arrays focus on those properties for ranking

**Review by gmarull — DISMISSED** [review]

**Review by hubertmis — DISMISSED** [review]

File: dts/bindings/clock/nordic,nrf54h-hfxo.yaml [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+    const: 32000000
+
+  accuracy-ppm:
+    type: int
+    description: Clock accuracy in parts per million
+    required: true
+
+  startup-time-us:
+    type: int
+    description: Clock startup time in micro seconds
```
[the remark this answers is not in the document]
Comment (hubertmis) [reply dts/bindings/clock/nordic,nrf54h-hfxo.yaml:21]: The BICR generator should take this value as an input.
@hakonfam, FYI

**Comment by bjarki-andreasen** [comment]
Added a DNM until I have added some patches we identified downstream :)

**Comment by bjarki-andreasen** [comment]
The latest changes modify Kconfig to not select NRF2_CLOCK_CONTROL by default, and removes the commit moving trimming of hsfll clock from the device driver init to the soc to ensure it is executed even if the device driver is not included (included by NRF2_CLOCK_CONTROL). This helps save flash in resource constrained applications.

**Comment by hubertmis** [comment]
> The latest changes modify Kconfig to not select NRF2_CLOCK_CONTROL by default, and removes the commit moving trimming of hsfll clock from the device driver init to the soc to ensure it is executed even if the device driver is not included (included by NRF2_CLOCK_CONTROL). This helps save flash in resource constrained applications.

Is it temporary or a final solution? I understood we aim to follow other vendors and keep clock-related initialization in the clock drivers. If we aim for resource-constrained applications, we should introduce Kconfigs to remove parts of the clock drivers. However, initialization should be obligatory if it's required for proper SoC operations.

**Comment by bjarki-andreasen** [comment]
> > The latest changes modify Kconfig to not select NRF2_CLOCK_CONTROL by default, and removes the commit moving trimming of hsfll clock from the device driver init to the soc to ensure it is executed even if the device driver is not included (included by NRF2_CLOCK_CONTROL). This helps save flash in resource constrained applications.
> 
> Is it temporary or a final solution? I understood we aim to follow other vendors and keep clock-related initialization in the clock drivers. If we aim for resource-constrained applications, we should introduce Kconfigs to remove parts of the clock drivers. However, initialization should be obligatory if it's required for proper SoC operations.

It is temporary :) clock init should definitely be moved to the clock driver once we figure out how optimize it for size :) I have created an internal ticket to track it :)

**Review by hubertmis — DISMISSED** [review]

File: drivers/clock_control/clock_control_nrf2_common.h [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+};
+
+/**
+ * @brief Defines a type for specific clock configuration structure.
+ *
+ * @param type suffix added clock_config_ to form the type name.
+ * @param _onoff_cnt number of clock configuration options to be handled;
+ *                   for each one a separate onoff manager instance is used.
+ */
+#define NRF2_STRUCT_CLOCK_CONFIG(type, _onoff_cnt) \
```
Comment (anangl) [inline drivers/clock_control/clock_control_nrf2_common.h:30]: This is not a public API, so I don't think we need to use these `nrf2_` prefixes here (which IMO look a bit weird without `clock_control_` before them).

**Comment by bjarki-andreasen** [comment]
Removed `nrf2_` namespace from private functions of the nrf2 clock control drivers, and adjusted licenses I had missed from `LicenseRef-Nordic-5-Clause` to `Apache-2.0`

File: dts/bindings/clock/nordic,nrf-lfclk.yaml [hunk 18]
```diff
@@ -0,0 +1,42 @@
+# Copyright (c) 2024 Nordic Semiconductor ASA
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
```
[the remark this answers is not in the document]
Comment (bjarki-andreasen) [reply dts/bindings/clock/nordic,nrf-lfclk.yaml]: done

File: dts/bindings/clock/nordic,nrf-fll16m.yaml [hunk 19]
```diff
@@ -0,0 +1,42 @@
+# Copyright (c) 2024 Nordic Semiconductor ASA
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
```
[the remark this answers is not in the document]
Comment (bjarki-andreasen) [reply dts/bindings/clock/nordic,nrf-fll16m.yaml]: done

File: drivers/clock_control/clock_control_nrf2_common.h (hunk 17 above)
Comment (bjarki-andreasen) [reply drivers/clock_control/clock_control_nrf2_common.h]: done

File: drivers/clock_control/clock_control_nrf2_hsfll.c [hunk 20]
```diff
@@ -0,0 +1,237 @@
...
+		return &dev_data->clk_cfg.onoff[0].mgr;
+	}
+
+	if (spec->accuracy || spec->precision) {
+		LOG_ERR("invalid specification of accuracy or precision");
+		return NULL;
+	}
+
+	frequency = spec->frequency == NRF_CLOCK_CONTROL_FREQUENCY_MAX
+		  ? MHZ(320)
```
Comment (anangl) [inline drivers/clock_control/clock_control_nrf2_hsfll.c:116]: This value is quite hidden here. It would be better to have it defined somewhere near the `clock_options` array.

File: samples/boards/nrf/nrf2_clock_control/Kconfig [hunk 21]
```diff
@@ -0,0 +1,24 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+config SAMPLE_CLOCK_FREQUENCY_HZ
+	int "Frequency specification to request from clock in Hz"
+
+config SAMPLE_CLOCK_ACCURACY_PPM
+	int "Accuracy specification to request from clock in PPM"
+
+config SAMPLE_CLOCK_PRECISION
+	int "Precision specification to request from clock"
```
Comment (anangl) [inline samples/boards/nrf/nrf2_clock_control/Kconfig:11]: Wouldn't it make sense to use `default 0` for these options (the API defines that 0 means: no specific requirement) and change them only when a certain value is required?

File: samples/boards/nrf/nrf2_clock_control/configs/fll16m.conf [hunk 22]
```diff
@@ -0,0 +1,8 @@
+# Copyright (c) 2024 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+CONFIG_CLOCK_CONTROL_NRF2=y
+
+CONFIG_SAMPLE_CLOCK_FREQUENCY_HZ=16000000
```
Comment (anangl) [inline samples/boards/nrf/nrf2_clock_control/configs/fll16m.conf:6]: Why request any value here when there is anyway no other choice than 16 MHz? Similarly for `lfclk`, it is always 32768. And for UART, users don't actually need to know/care what is the frequency of the clock that drives it.

File: drivers/clock_control/clock_control_nrf2_common.c [hunk 23]
```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "clock_control_nrf2_common.h"
+#include <zephyr/kernel.h>
```
[the remark this answers is not in the document]
Comment (bjarki-andreasen) [reply drivers/clock_control/clock_control_nrf2_common.c]: removed #include <zephyr/kernel.h>

File: samples/boards/nrf/nrf2_clock_control/src/main.c [hunk 24]
```diff
@@ -0,0 +1,98 @@
...
+
+#define SAMPLE_NOTIFY_TIMEOUT       K_SECONDS(2)
+#define SAMPLE_PRE_REQUEST_TIMEOUT  K_SECONDS(CONFIG_SAMPLE_PRE_REQUEST_TIMEOUT)
+#define SAMPLE_KEEP_REQUEST_TIMEOUT K_SECONDS(CONFIG_SAMPLE_KEEP_REQUEST_TIMEOUT)
+
+const struct device *sample_clock_dev = DEVICE_DT_GET(SAMPLE_CLOCK_NODE);
+
+static K_SEM_DEFINE(sample_sem, 0, 1);
+
+static void sample_notify_cb(void)
```
[the remark this answers is not in the document]
Comment (bjarki-andreasen) [reply samples/boards/nrf/nrf2_clock_control/src/main.c]: updated signature

File: drivers/clock_control/clock_control_nrf2_hsfll.c (hunk 20 above)
Comment (bjarki-andreasen) [reply drivers/clock_control/clock_control_nrf2_hsfll.c]: added defines
```
#define HSFLL_FREQ_LOW    MHZ(64)
#define HSFLL_FREQ_MEDLOW MHZ(128)
#define HSFLL_FREQ_HIGH   MHZ(320)
```
and used this across file

File: samples/boards/nrf/nrf2_clock_control/Kconfig (hunk 21 above)
Comment (bjarki-andreasen) [reply samples/boards/nrf/nrf2_clock_control/Kconfig]: added default 0 along with help to indicate what values of 0, 1 and >1 means

File: samples/boards/nrf/nrf2_clock_control/configs/fll16m.conf (hunk 22 above)
Comment (bjarki-andreasen) [reply samples/boards/nrf/nrf2_clock_control/configs/fll16m.conf]: removed frequency (now uses default 0)

File: samples/boards/nrf/nrf2_clock_control/CMakeLists.txt
Comment (anangl) [inline samples/boards/nrf/nrf2_clock_control/CMakeLists.txt:1]: `nrf2_clock_control` does not seem to be a good name for this sample. I'd go with just `clock_control`.

File: tests/drivers/clock_control/nrf2_clock_control/CMakeLists.txt
Comment (anangl) [inline tests/drivers/clock_control/nrf2_clock_control/CMakeLists.txt:1]: Similarly here, instead of `nrf2_clock_control`, I'd use `nrf_clock_control`.

File: samples/boards/nrf/nrf2_clock_control/Kconfig [hunk 25]
```diff
@@ -0,0 +1,38 @@
+# Copyright (c) 2024 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+config SAMPLE_CLOCK_FREQUENCY_HZ
+	int "Frequency specification to request from clock in Hz"
+	default 0
+	help
+	  0  -> ignore frequency
+	  1  -> use max frequency
+	  >1 -> use at minimum selected frequency
```
[the remark this answers is not in the document]
Comment (bjarki-andreasen) [reply samples/boards/nrf/nrf2_clock_control/Kconfig]: Updated to state:
```
	  >1 -> use at minimum selected frequency. To select the
	        highest supported frequency use UINT32_MAX.
```

File: samples/boards/nrf/nrf2_clock_control/CMakeLists.txt
Comment (bjarki-andreasen) [reply samples/boards/nrf/nrf2_clock_control/CMakeLists.txt:1]: renamed to clock_control

File: tests/drivers/clock_control/nrf2_clock_control/CMakeLists.txt
Comment (bjarki-andreasen) [reply tests/drivers/clock_control/nrf2_clock_control/CMakeLists.txt:1]: renamed to nrf_clock_control

File: samples/boards/nrf/nrf2_clock_control/Kconfig (hunk 25 above)
[the remark this answers is not in the document]
Comment (anangl) [reply samples/boards/nrf/nrf2_clock_control/Kconfig]: It should say `>0`, actually.

**Comment by hubertmis** [comment]
Temporarily, we should not use closed-loop modes. Should we update this PR or remove the closed-loop later?

**Comment by bjarki-andreasen** [comment]
> Temporarily, we should not use closed-loop modes. Should we update this PR or remove the closed-loop later?

I would like to remove closed loop later, to keep consistency between the commits in this PR and the downstream PR (which is already merged)

File: samples/boards/nrf/nrf2_clock_control/Kconfig (hunk 25 above)
[the remark this answers is not in the document]
Comment (bjarki-andreasen) [reply samples/boards/nrf/nrf2_clock_control/Kconfig]: Updated to >0

**Review by anangl — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Review by nordic-krch — APPROVED** [review]
